### DISCUSSION TITLE
Add config for oauth1

### DISCUSSION THREAD

**Discussion by jdevosMagento** [apideck-libraries/portman#270] [Q&A]
Hello,

I need to test an API using oauth1, to be configured with Consumer Key, Consumer Secret, Access Token and Token Secret.

How can I configure portman using this method?

Thank you

**Comment by thim81** [comment]
hi @jdevosMagento 

Are you using the Oauth Flows provided by Postman or doing it directly as part of a Pre-request script or want to validate the API endpoints that are part of your collection?

If it is a pre-request script, you could have a look at this [GIST](https://gist.github.com/thim81/a1e218272fbcf934bcbc2d19cab70fef) that I use for getting a token from Auth0. 
It is based on a Postman Environment that contains some Oauth variables and the pre-request part, that fetches the token and stores that as variable.

**Reply by jdevosMagento** [reply]
I use what Postman offers 

![capture_postman](https://user-images.githubusercontent.com/63316093/157637626-52d7bbe8-2114-43cb-9f12-d853d83097ca.png)

**Reply by thim81** [reply]
The Oauth flows in Postman are quite special and I never really got them to work properly (hence my approach of using Pre-request scripts which gives me full freedom to implement).

In Portman we have support for the methods: apiKey, bearer, basic
https://github.com/apideck-libraries/portman/tree/main/examples/portman-globals#securityoverwrites

![2022-03-10 at 11 28 19@2x](https://user-images.githubusercontent.com/952446/157643041-87896b6c-ba87-4f2b-97a9-596f4ab1d8a6.png)

Perhaps if you could provide us a Postman collection with the expected Oauth1.0 data filled in, we can investigate to extend the "securityOverwrites" with the Oauth 1.0 method?

**Reply by jdevosMagento** [reply]
I see that there is an "other" possibility for securityOverwrites 
https://github.com/apideck-libraries/portman/blob/4c715f191a37f21649ab8cbc2ad4123f8985cba1/src/utils/portman-config-schema.json#L165

But I can't get it to work for the moment, the export to postman does not return any information.

With the pre-request script, the portman-cli with newman testing its possible/OK ?

**Reply by thim81** [reply]
The JSON schema is generated so I expect that the "other" option is not implemented but the result of an unused variable that still remained.

That is also why Portman will return any information, since nothing is happening with the defined "other".

I can confirm that the pre-request approach works in Postman & Newman, since we use to test our API contracts using the Auth2 pre-request script.

FYI: I found an OAuth1 example, which I'll review to see if we can introduce as security overwrite. I created a [work item](https://github.com/apideck-libraries/portman/issues/271) for it.

**Reply by jdevosMagento** [reply]
It's ok with the "overwriteRequestSecurity" and "other" for sending the data to postman, and newman send a 200

![capture_oauth1](https://user-images.githubusercontent.com/63316093/157650553-db21bac6-6674-4815-a171-e86dde7d77f8.png)

![capture_oauth1_2](https://user-images.githubusercontent.com/63316093/157650577-f1c4ebe0-81a2-4a73-8a71-f56d926ada61.png)

But in "generals" it is not recognized, have to put in "overwrites"

**Reply by thim81** [reply]
Thanks for the input, we will review the option to bring the security overwrites on a collection level.

**Comment by thim81** [comment]
@jdevosMagento Would you be OK if we provide the globals options for securityOverwrites, where you would insert the Postman configuration AS IS defined in the collection?

Example:

```json
{
  "version": 1.0,
  "globals": {
    "securityOverwrites": {
      "oauth1": [
        {
          "key": "addEmptyParamsToSign",
          "value": true,
          "type": "boolean"
        },
        {
          "key": "timestamp",
          "value": "1461319769",
          "type": "string"
        },
        {
          "key": "nonce",
          "value": "ik3oT5",
          "type": "string"
        },
        {
          "key": "consumerSecret",
          "value": "D+EdQ-gs$-%@2Nu7",
          "type": "string"
        },
        {
          "key": "consumerKey",
          "value": "RKCGzna7bv9YD57c",
          "type": "string"
        },
        {
          "key": "signatureMethod",
          "value": "HMAC-SHA1",
          "type": "string"
        },
        {
          "key": "version",
          "value": "1.0",
          "type": "string"
        },
        {
          "key": "addParamsToHeader",
          "value": false,
          "type": "boolean"
        }
      ]
    }
  }
}
```

**Reply by jdevosMagento** [reply]
yes it's great like that! This will avoid duplicating accesses for each path

**Reply by thim81** [reply]
This would be an as-is extract from the Postman collection (even including the "type"). That makes sense for you?

**Reply by thim81** [reply]
We have a [PR](https://github.com/apideck-libraries/portman/pull/272) in the making, so this might land in the next release of Portman.

**Reply by jdevosMagento** [reply]
I think the type is not useful, there in the "overwriteRequestSecurity" part, there is only key: value and postman in its collection regenerates the type itself.

**Reply by thim81** [reply]
I guess it depends what would be the typical way to define it: manually or copy pasting it from the Postman collection export?

**Reply by jdevosMagento** [reply]
Actually, with overwriteRequestSecurity part, the collection is generated by portman : 

"auth": {
              "type": "oauth1",
              "oauth1": [
                {
                  "type": "any",
                  "value": "xxx",
                  "key": "consumerKey"
                },
                {
                  "type": "any",
                  "value": "xxx",
                  "key": "consumerSecret"
                },
                {
                  "type": "any",
                  "value": "xxx",
                  "key": "token"
                },
                {
                  "type": "any",
                  "value": "xxx",
                  "key": "tokenSecret"
                },
                {
                  "type": "any",
                  "value": "HMAC-SHA1",
                  "key": "signatureMethod"
                },
                {
                  "type": "any",
                  "value": true,
                  "key": "addParamsToHeader"
                }
              ]
            }
          },

After inserted this on postman, i export manually on postman for check and voila : 

"request": {
						"auth": {
							"type": "oauth1",
							"oauth1": [
								{
									"key": "consumerKey",
									"value": "xxxx",
									"type": "string"
								},
								{
									"key": "consumerSecret",
									"value": "xxxx",
									"type": "string"
								},
								{
									"key": "token",
									"value": "xxxx",
									"type": "string"
								},
								{
									"key": "tokenSecret",
									"value": "xxxx",
									"type": "string"
								}
							]
						},

**Reply by thim81** [reply]
I'm a bit confused now?

If you set the `overwriteRequestSecurity` and you export the collection, does it set the Authentication on the Postman collection level?

**Reply by jdevosMagento** [reply]
No it sets it only on the path, not on the whole collection.

**Reply by thim81** [reply]
Ok, clear. The PR provides the securityOverwrite on Postman collection level. Still remaining item is if the config should be provided as the full Postman syntax: key/value/type OR only key/value

**Comment by thim81** [comment]
Since Portman 1.15.0, we introduced the option to pass the Postman authentication options.
For more information, please review the [securityoverwrites](https://github.com/apideck-libraries/portman/blob/main/examples/portman-globals/readme.md#securityoverwrites) example.
